File: src/main/org/nlogo/prim/etc/_randomnormal.java
package org.nlogo.prim.etc;

import org.nlogo.api.LogoException;
import org.nlogo.nvm.Context;
import org.nlogo.nvm.EngineException;
import org.nlogo.nvm.Reporter;
import org.nlogo.nvm.Syntax;

public final strictfp class _randomnormal extends Reporter {
  @Override
  public Syntax syntax() {
    int[] right = {Syntax.TYPE_NUMBER, Syntax.TYPE_NUMBER};
    return Syntax.reporterSyntax(right, Syntax.TYPE_NUMBER);
  }

  @Override
  public Object report(Context context) throws LogoException {
    return report_1(context,
        argEvalDoubleValue(context, 0),
        argEvalDoubleValue(context, 1));
  }

  public double report_1(Context context, double mean, double sdev) throws LogoException {
    if (sdev < 0) {
      throw new EngineException(
          context, this,
          "random-normal's second input can't be negative");
    }
    return validDouble(mean + sdev * context.job.random.nextGaussian());
  }
}

File: src/main/org/nlogo/prim/etc/_randomnormal.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim.etc

import org.nlogo.nvm.{ Context, EngineException, Reporter }
import org.nlogo.api.{ I18N, Syntax }

class _randomnormal extends Reporter {
  override def syntax =
    Syntax.reporterSyntax(Array(Syntax.NumberType,
                                Syntax.NumberType),
                          Syntax.NumberType)
  override def report(context: Context) =
    Double.box(
      report_1(context,
               argEvalDoubleValue(context, 0),
               argEvalDoubleValue(context, 1)))
  def report_1(context: Context, mean: Double, sdev: Double): Double = {
    if (sdev < 0.0)
      throw new EngineException(
          context, this, I18N.errors.get(
            "org.nlogo.prim.etc._randomNormal.secondInputNotNegative"))
    validDouble(mean + sdev * context.job.random.nextGaussian)
  }
}
